tutorials/Tutorial18_GPL.py

```diff
@@ -0,0 +1,249 @@
+        docs = [
+            "Corona is transmitted via the air",
+            "Ebola is transmitted via direct contact with blood",
+            "HIV is transmitted via sex or sharing needles",
+            "Polio is transmitted via contaminated water or food",
+        ]
+
+        query_emb = model.encode(query)
+        docs_emb = model.encode(docs)
+        scores = util.dot_score(query_emb, docs_emb)[0]
+        doc_scores = sorted(zip(docs, scores), key=lambda x: x[1], reverse=True)
+
+        print("Query:", query)
+        for doc, score in doc_scores:
+            # print(doc, score)
+            print(f"{score:0.02f}\t{doc}")
+
+    print("Original Model")
+    show_examples(org_model)
+
+    """# Get Some Data on COVID-19
+    We select 10k scientific publications (title + abstract) that are connected to COVID-19. As dataset we use [TREC-COVID-19](https://huggingface.co/datasets/nreimers/trec-covid).
+    """
+
+    dataset = load_dataset("nreimers/trec-covid", split="train")
+    num_documents = 10000
+    corpus = []
+    for row in dataset:
+        if len(row["title"]) > 20 and len(row["text"]) > 100:
+            text = row["title"] + " " + row["text"]
+
+            text_lower = text.lower()
+
+            # The dataset also contains many papers on other diseases. To make the training in this demo
+            # more efficient, we focus on papers that talk about COVID.
+            if "covid" in text_lower or "corona" in text_lower or "sars-cov-2" in text_lower:
+                corpus.append(text)
+
+            if len(corpus) >= num_documents:
+                break
+
+    print("Len Corpus:", len(corpus))
+
+    """# Initialize Haystack Retriever and DocumentStore
+    
+    Let's add corpus documents to `FAISSDocumentStore` and update corpus embeddings via `EmbeddingRetriever`
+    """
+
+    from haystack.nodes.retriever import EmbeddingRetriever
+    from haystack.document_stores import FAISSDocumentStore
+
+    document_store = FAISSDocumentStore(faiss_index_factory_str="Flat", similarity="cosine")
+    document_store.write_documents([{"content": t} for t in corpus])
+
+    retriever = EmbeddingRetriever(
+        document_store=document_store,
+        embedding_model="sentence-transformers/msmarco-distilbert-base-tas-b",
+        model_format="sentence_transformers",
+        max_seq_len=max_seq_length,
+        progress_bar=False,
+    )
+    document_store.update_embeddings(retriever)
+
+    """## Optionally download pre-generated questions or even generate them outside of Haystack
+    
+    The first step of the GPL algorithm requires us to generate questions for a given text passage. Even though our pre-COVID trained model hasn't seen any COVID-related content, it can still produce sensible queries by copying words from the input text. As generating questions from 10k documents is a bit slow (depending on GPU used), we'll download question/document pairs directly from the HuggingFace hub.
```

Comment:
```suggestion
    The first step of the GPL algorithm requires us to generate questions for a given text passage. Even though our pre-COVID trained model hasn't seen any COVID-related content, it can still produce sensible queries by copying words from the input text. As generating questions from 10k documents is a bit slow (depending on the GPU used), we'll download question/document pairs directly from the Hugging Face hub.
```